### DISCUSSION TITLE
[UI] Custom Blender Interfaces and Data Passthrough

### DISCUSSION THREAD

**Discussion by panthavma** [bnpr/Malt#481] [Development]
Hello, I have a more technical discussion to start, for solutions to get more advanced custom tools in.

### Problem Statement
So at the moment Malt allows you to define custom pipelines which can do quite a lot, which is cool and can enable more advanced tools. The current issue I see is that the automatic interface generation can be a bit limited at times, for instance when you have a lot of materials / parameters at once (which then becomes hard to navigate and can even slow down the software pretty hard when reaching higher numbers). It also makes it a bit technical at times, by having raw variables to tune, which can make it unsuitable for less-technical artists.

I think adding some way to be able to communicate both ways with that interface could help make more complex tools and pipelines.

### Example Use cases
- Having buttons for starting longer offline processes
- Getting text feedback from the custom pipeline (like profiling or other internal parts)
- Selecting which inputs to show, or providing toggles / categories that can slim down
- Custom widgets for things like toon ramps or spicy BRDFs
- Pass information like selected meshes or vertices

### Potential Issues
- Malt is made to be independent of Blender by design (with the BlenderMalt + Bridge), which means that this feature would go a bit against this part. I don't think it's impossible to reconciliate the two, but might require more indirect solutions.
- It might make custom pipelines very busy and have to manage several roles, or make the interface complicated, or make it hold a lot of data that might be unnecessary. (depends on solution)
- It might introduce more disparity between custom pipelines and nodal pipelines, but I think it's okay since it would be for complex applications that would be impractical with just nodal tools imo. 

### Implementation / Design Ideas
The global solution ideas I may envision for that issue (first draft, probably missing a lot of constraints):
1. One that would require no change to Malt would be creating a second module with access to bpy, and make them communicate via sockets or a similar IPC mechanism. While this would be functional, it would be pretty finicky (what happens when Malt restarts / crashes / has a compilation error / isn't started, provides access twice to the same parameters while not solving the overcrowding issue), is a lot more complex to put in practice, and makes setup harder. This also would not allow access to the graphics pipeline, and thus complicate some workflows that would rely on visible objets or compute shaders I think.
2. Could add more parameters to WorldParameters / Materials / etc and callbacks, but this would be pretty messy and break that separation between Malt and Blender.
3. Could mirror the custom pipeline functionality with a "Custom Interface" file. I might be a bit heavy codewise, but I think it might be the best option for good integration with Malt, while keeping the separation intact.

I'll annotate that I'm not an expert with Blender / Python / bpy itself so there might be quite a few elements that I missed. This is also not a "straight" up request, I'm looking for a way to implement it and contribute it back in a way that works with Malt's design, and need guidance for that.

Anyway, please let me know what you think, what do you think I missed, what could be improved, potentially other use cases that could affect design! I'm pinging @pragma37 also since he's the Malt expert and we talked about passing some additional data down in the past (but not to that extent).

**Comment by pragma37** [comment]
Some nice ideas here. 😀

In case you don't know, the plugin system already lets you register stuff on the Blender side:
https://github.com/bnpr/Malt/blob/Development/Malt/PipelinePlugin.py#L22
We could add callbacks for plugins in BlenderMalt if needed.

While it's not currently supported, ideally, it should be possible to register custom pipelines from plugins too.
That way anything related to a custom pipeline could be packed/distributed together.

I haven't thought of this before, but I think a custom message system for the Bridge API would be a nice addition.
You send an arbitrary message from Blender and it would be up to the pipeline to handle it.
We could support sending data back too, but that would require either blocking Blender or implementing a poll system.

This would allow implementing buttons, but also sending extra data to Malt.
Actually, it might even make sense to have a built-in Button Parameter Type to register callbacks directly from the pipeline.

> Getting text feedback from the custom pipeline (like profiling or other internal parts)

There's already a shared dictionary (Bridge.shared_dict), with an entry used for printing performance stats in Blender.
We could expose it to the pipeline side too. You could also add custom entries from Blender and write data to it from the render pipeline.

The trickiest part would be the custom UI stuff for BlenderMalt parameters.
As they're now, it's a leave it or take it kind of deal, and the system is already quite complex as it is.
Also, the Blender UI system is quite restrictive, there's no support for custom widgets or anything like that.

If you have specific performance or UX issues we can also try to solve them at the core, instead of through plugins, though.

**Comment by panthavma** [comment]
> In case you don't know, the plugin system already lets you register stuff on the Blender side:
Oh I didn't see that addition, nice :)

Yeah, a button type would be very useful, and a simple message API would work pretty well. I don't see how to implement them exactly here though.

I think my UX issues come into two camps:
- Too many parameters with too many visible things. One of my pipelines ideally needs to be able to handle showing 1000 materials, which is slow, but could be doable if there is a system to hide them. I saw that malt remembers the values, for undeclared parameters, but I think some way to force a refresh and/or show which ones to hide or group would be good.
- Really advanced parts, like needing to select a light to use as a source to tweak and load parameters on the model, which then get rendered, in order to make advanced tools (for research atm). I'm not sure it's something that can or should be solved on the malt only side.

I don't know if you have ideas for those ? That might be independent of this discussion or linked to it, and might be very specific to my needs (needing a centralized place to hold and bake data, and needing access to the GPU AND a UI that's flexible enough for experiments).

**Comment by pragma37** [comment]
> Yeah, a button type would be very useful, and a simple message API would work pretty well. I don't see how to implement them exactly here though.

Something along these lines (I haven't tested it):
[custom_msg.patch](https://github.com/bnpr/Malt/files/10761084/custom_msg.patch)
I'm not sure if it would make sense to call the callback for each active viewport, and maybe pass the scene along. 

> Too many parameters with too many visible things. One of my pipelines ideally needs to be able to handle showing 1000 materials, which is slow, but could be doable if there is a system to hide them.

I'm not sure what you mean here. Are still you talking about the UI?

> I saw that malt remembers the values, for undeclared parameters, but I think some way to force a refresh and/or show which ones to hide or group would be good.

Yep, this has been on my to-do since forever. It shouldn't be too hard.
It's more about figuring out the UX.

> Really advanced parts, like needing to select a light to use as a source to tweak and load parameters on the model, which then get rendered, in order to make advanced tools (for research atm). I'm not sure it's something that can or should be solved on the malt only side.

That sounds like something that would have to be done on the Blender side.

**Comment by panthavma** [comment]
> Something along these lines (I haven't tested it):

Oh I see, would be useful indeed!

> I'm not sure if it would make sense to call the callback for each active viewport, and maybe pass the scene along.

Hmmm... I think it makes more sense to only have it being called once, since you only press the button once and there's a single Malt instance. If we have a method to force a refresh of the current viewports, I think it will cover most usecases. Passing the scene seems useful.

> I'm not sure what you mean here. Are still you talking about the UI?

Yes, I meant literally showing the parameters is slowing down my Blender lol, the shader / pipeline is going fast, but scrolling the material list is unpractical and very laggy (bordering on a second between scroll refreshes).

> Yep, this has been on my to-do since forever. It shouldn't be too hard.
> It's more about figuring out the UX.

What are the parts we should figure out ? UX for the programmers, or UX for the users ?
For the programmers, I think you can add some additional arguments / parameters to the Parameter class. For the users, I'm not exactly sure, but I think having parameters grouped together is already a good start.

Maybe the simplest solution is having each Parameter have a `category`, which can be null or a string, and simply grouping them based on that. Feels simple enough to implement on both sides, easy to add for the programmers, and pretty transparent for users. I'm not sure how it interacts with nodes however, and it's just part of a solution (albeit a quick win already I think). If you have other ideas I'm all ears :)

> That sounds like something that would have to be done on the Blender side.

Yep that probably is something way too specific / involved. The plugin system you mentioned should be able to help there from what I understand ?
I'm a bit confused with the interface, it seems like you can load plugins from the current directory automatically, but what exactly do they do (ie, is that a self contained plugin, or a way to connect an existing one to malt?) ? Since it's in Malt, do they have access to bpy ? Is there an example I can look up ?

---

I feel like the discussion is moving along three points:

- Buttons / messages
- User interface for parameters
- Plugins

Not to rush anything along since we're still discussing, but I think structure is useful over text, and we might be able to "convert" them into actual tasks later.

**Comment by pragma37** [comment]
> I think it makes more sense to only have it being called once, since you only press the button once and there's a single Malt instance. If we have a method to force a refresh of the current viewports, I think it will cover most usecases. Passing the scene seems useful.

The issue is that each Viewport has its own Scene and Pipeline instance, so if you want to do something scene-related you would have to do it per Viewport.

I think we might be able to share the same scenes across viewports as long as they use the same Scene and ViewLayer on the Blender side (except for the Camera).

> Yes, I meant literally showing the parameters is slowing down my Blender lol, the shader / pipeline is going fast, but scrolling the material list is unpractical and very laggy (bordering on a second between scroll refreshes).

So you are pre-registering from your Pipeline 1000 Parameters of type Material? I guess in the World or Scene tab?
I've experimented with adding folding to the UI parameters before. I could give it another look, but even then I'm not sure if that would be enough...
You might try another approach, like adding some kind of special object on the Blender side with 1000 material slots and giving them pre-defined names, so you can find them on the Pipeline side.

> What are the parts we should figure out ? UX for the programmers, or UX for the users ?

For users, the MaltProperties class already has a remove_property function.
Adding the feature to the UI in a way that is actually useful, unintrusive, and safe for the user is more involved than what you might expect.

> The plugin system you mentioned should be able to help there from what I understand ?
I'm a bit confused with the interface, it seems like you can load plugins from the current directory automatically, but what exactly do they do (ie, is that a self contained plugin, or a way to connect an existing one to malt?) ?

You can think of blendermalt_register/unregister like the regular register/unregister functions all Blender addons use.
Their only advantage is that you can be sure that BlenderMalt has already been registered when they are called, which is not the case for a regular Blender addon.
https://github.com/bnpr/Malt/blob/Development/BlenderMalt/__init__.py#L221

So yep, you have full access to the bpy API.
On the other hand, you don't have direct access to the Malt Server/Pipeline.

> Not to rush anything along since we're still discussing, but I think structure is useful over text, and we might be able to "convert" them into actual tasks later.

I think pull requests would work just fine?

**Comment by panthavma** [comment]
> The issue is that each Viewport has its own Scene and Pipeline instance, so if you want to do something scene-related you would have to do it per Viewport.
> I think we might be able to share the same scenes across viewports as long as they use the same Scene and ViewLayer on the Blender side (except for the Camera).

Oh I didn't know that. I guess you can also change the displayed objects depending on the exact viewport ? That could be problematic if you're doing something like writing to the disk, if two processes try the same file at the same time. Maybe we could add a flag to the first one called (or an ID to each) so that it's sure to not mix up between processes ?

> So you are pre-registering from your Pipeline 1000 Parameters of type Material? I guess in the World or Scene tab?
I've experimented with adding folding to the UI parameters before. I could give it another look, but even then I'm not sure if that would be enough...

Yep, from the world parameters. I'm not sure where else I can get something I can reliably get back on the Malt side. I don't necessarily need all of them at the same time, oftentimes just showing 16 would be enough, but I can't trigger the refresh effectively without relaunching the malt pipeline.

> You might try another approach, like adding some kind of special object on the Blender side with 1000 material slots and giving them pre-defined names, so you can find them on the Pipeline side.

That could be a solution (albeit more finicky I think), but I'm a bit confused: you can get the objects by name ? Last time we talked about this iirc you said it would be something that's possible, but not fit for general use (since it was in the context of mesh loading and could very well be a different mesh).

> For users, the MaltProperties class already has a remove_property function.
Adding the feature to the UI in a way that is actually useful, unintrusive, and safe for the user is more involved than what you might expect.

I'll take a look at remove_property, but I'm not sure exactly that it should be something for the end-user.
The way I see it, for these pipelines, there's going to be two roles for the tools made in Malt:
- The creator: the person who creates the tool using malt, either with programming or nodes, and exposes parameters
- The end user: the person using the tool and specifying the parameters' values

They can be the same, but I think the UI manipulation features should stay in the creator part, since they are the ones knowledgeable about their tool. End users in general are less technical, and leaving the option open is gonna be a source of errors and make it more complex for them. Based on the experience I had working it's better to keep it simple for them.

Linking UI display options to the parameters make sense imo, because:
- As a user, you interact often by changing parameters so it makes sense to have them specify how they are displayed.
- As a creator/programmer, you can define and modify them easily, so no problem here.
- As a creator/nodal, you can probably display a fair amount of them as parameters to those parameter nodes.
- TBC, but the parameters are already iterated on by BlenderMalt, therefore having them here doesn't require a new pathway.

I think for starters, we can add two display options that I think are useful and I can think of:
- Hidden: it isn't displayed. This ensures it stays saved with the pipeline. If combined with a way to request refreshes, it should offer a more in depth interface, even if a bit hacky.
- Category: just a string for a category, all parameters with the same one will be rendered together in a foldable category.

I think this should already cover a fair amount of ground to start out, while staying within the current codebase as I understand it and fitting with common use case patterns, so I think it's worth implementing. What do you think ?

> So yep, you have full access to the bpy API.
On the other hand, you don't have direct access to the Malt Server/Pipeline.

Thanks! I'll check it out. How do you think would be best to have that link to the Malt server/pipeline ? Could work with some IPC, but I think modules registered this way can ignore the Blender / Malt separation by design of being modules requiring both Malt and Blender, so having a built-in way might be better ? I don't know python and blender enough to think of repercussions of just giving the open pipelines as a parameter or something.

> I think pull requests would work just fine?

Yeah, I meant we can probably separate them into concrete tasks with solutions to implement when we decide on them, so that we can then merge them as pull requests. :)
Don't want to put additional work on your plate here, just to find what would work best for malt so that it's not just a one-sided addition.

**Comment by pragma37** [comment]
> Oh I didn't know that. I guess you can also change the displayed objects depending on the exact viewport ? That could be problematic if you're doing something like writing to the disk, if two processes try the same file at the same time. Maybe we could add a flag to the first one called (or an ID to each) so that it's sure to not mix up between processes ?

There's only one Malt process for each Blender instance, different viewports are rendered sequentially.

> That could be a solution (albeit more finicky I think), but I'm a bit confused: you can get the objects by name ? Last time we talked about this iirc you said it would be something that's possible, but not fit for general use (since it was in the context of mesh loading and could very well be a different mesh).

The object's names are not sent, but all the resources (meshes, textures, materials) are loaded through a proxy system and they are all stored in the scene in a dictionary, so you should be able to retrieve them with `scene.proxys[('material', material_name)]`.

Maybe you could instead add some enum or string parameters to the material themselves, so you can search for the relevant ones?
It's hard to give you a better answer without knowing any details on how these materials are used or why you need 1000.

> I think for starters, we can add two display options that I think are useful and I can think of:
>
>    Hidden: it isn't displayed. This ensures it stays saved with the pipeline. If combined with a way to request refreshes, it should offer a more in depth interface, even if a bit hacky.
    Category: just a string for a category, all parameters with the same one will be rendered together in a foldable category.

All caps parameters and parameters with a leading underscore are already hidden in the UI.
I don't think adding a UI option so the user can hide individual parameters makes too much sense.
Node-based materials already have a system for selecting which parameters are shown in the material UI.

We already have "categories" that get grouped together, .ie ` Viewport.Resolution Scale` and `Viewport.Smooth Interpolation`, the missing feature is being able to fold them in the UI.

> How do you think would be best to have that link to the Malt server/pipeline ? 

Through the Bridge interface, that's the only possible way. That's why I suggested the CUSTOM message type.
If you need to send huge amounts of data take a look at Bridge.get_shared_buffer(), otherwise regular Python objects are fine.

>  I think modules registered this way can ignore the Blender / Malt separation by design of being modules requiring both Malt and Blender, so having a built-in way might be better ?

You can certainly import a python module from both processes, and you can import any Malt module from BlenderMalt.
But it won't be THE SAME module, it's the same module loaded twice from different processes. 

> Yeah, I meant we can probably separate them into concrete tasks with solutions to implement when we decide on them, so that we can then merge them as pull requests. :)

Sure, what I mean is that it's ok to open a pull request for a specific feature even if the code is not ready to be merged, or even if there's no code yet at all but you want to discuss the implementation and design details.

**Comment by panthavma** [comment]
> There's only one Malt process for each Blender instance, different viewports are rendered sequentially.

Oh neat, that's one less problem to worry about. Then why would we want to call a function several times when pressing a button if there is only one instance ? I would think a single call is the simplest and expected behavior

---

> The object's names are not sent, but all the resources (meshes, textures, materials) are loaded through a proxy system and they are all stored in the scene in a dictionary, so you should be able to retrieve them with scene.proxys[('material', material_name)].

Oh I see! Thanks for the tip

> Maybe you could instead add some enum or string parameters to the material themselves, so you can search for the relevant ones?
It's hard to give you a better answer without knowing any details on how these materials are used or why you need 1000.

I didn't want to clog it with potentially unneeded info because it's a bit long haha, I'll try to keep it concise.

- My use case for this is to render for video games, using Godot 3, which is unfortunately limited to a single output image, but you can trick it into doing passes. I thus need to store all my model/material information AND lighting in the first pass, so instead I use IDs as a proxy, and get my material information from a palette texture in the second pass, similar to how we had indexed colors for images. Not ideal, but with some management its doable.
- The issue here is that I want to create those palette textures in Blender directly, so that I can work in my DCC and get the same render in engine to iterate faster. I do that by having an internal pass in Malt generate that texture and then either export it to disk, or use it directly for its own render. It's a simple pass, all the materials render to a quad sequentially, position themselves correctly on the texture, and output the needed parameters.
- There are a few constraints making it unsuitable for regular object parameters:
    - The materials can be shared between different objects, so if each one had it I could get into race conditions at the time of render.
    - Because of the limited bit size and quality needed, I use ID-Based lines, so I need to be able to reuse materials in different slots, which means a material doesn't know its own slot: it's supplied by the pipeline at rendertime, and can be called several times.
    - The information quantity for each material is also pretty big (depending on the style, it can be several toon ramps), so I can't really do the material sharing manually. Similarly, in order to set information quickly for the first pass, I don't use textures but instead hijack UV and set it manually.
- Therefore, all those materials are in WorldParameters, and it works really well for my purposes overall, and is fast enough with 256 materials. But, since I'm starting to share a lot of assets (for instance, on my robot fighting game, I have additional weapon packs that can't fit all together in the same palette texture, and create a palette for each predefined combi), I started to have a need for additional slots, either as additional palette textures, or as a way to alter a base palette texture, and 1024 materials is straining the UI too much (hard to navigate, but also very slow).

Here is a picture to help convey the idea, IDs are arbitrary:

![image](https://user-images.githubusercontent.com/22663538/219937388-dec498d1-1868-4cab-abfc-dd2d0271970c.png)

So my main issue here is improving the UX for me, so the solutions I imagined / tried:
1. I can choose which parameters I declare in the pipeline as world parameters, and thus its the only ones Blender shows. It seems to work and remember the parameters, but it requires reloading the pipeline each time (potentially by fiddling the python code since I'm not sure I'll have access to parameter values then), and not having them declared scares me a bit that Blender won't save them and I'll have to manually reassign them to their slots
2. I think the usual method of using all caps parameters to hide them wouldn't work, since they would point to different points of the dictionary
3. Implement a way to fold / hide parameters and/or force a refresh, which would require going in the Malt code and adding it, and seems like it would be a useful reusable feature.
4. Maybe with `remove_property` ? I haven't evaluated this yet because I learnt about it when you told me lol, but it seems like it would also encounter some of the problems of 1. (sans the risk of not saving)

So that was one of the objectives when making the post, to see if we could add some kind of way to categorize parameters as it would improve that UX for my case, but also in plenty of other cases by grouping parameters.

Hopefully that wasn't too horrifying of a pipeline lol, I'm self taught on Blender so I don't know the conventions, and by being able to program I have enough power to break stuff. If you think I missed something in the solutions, don't hesitate to let me know!

(to note, this pipeline isn't the only thing I came about here, I was wondering about the UX and plugin communication in general, since I know that I'll need that info in the future, but this is the most "practical" part).

> All caps parameters and parameters with a leading underscore are already hidden in the UI.
> I don't think adding a UI option so the user can hide individual parameters makes too much sense.

Yeah, I wasn't suggesting that. I think that default behavior is fine, maybe slightly better if you could get that control, but it's honestly not a big enough issue at the moment.

There's some other subjects in a similar vein for better integration with non-technical artists, but I think we can leave them for later lol.

> We already have "categories" that get grouped together, .ie  Viewport.Resolution Scale and Viewport.Smooth Interpolation, the missing feature is being able to fold them in the UI.

Okay I see, so basically, "just" adding the fold to that system would work well.
Is that grouping exposed, and so how ? From what I see in `Pipeline.py`, you just declare them with a dot, so I could do `GroupA.ParamX` and `GroupA.ParamY`, and they would get grouped together automatically ?

---

> Through the Bridge interface, that's the only possible way. That's why I suggested the CUSTOM message type.
If you need to send huge amounts of data take a look at Bridge.get_shared_buffer(), otherwise regular Python objects are fine.

Okay, so I'll try to phrase how I understand it works, let me know if I have a mistake somewhere or if you have more detail

- I have MaltPipeline, which is a malt custom pipeline, and CoolPlugin which is my Blender plugin. I want both to be able to communicate together. CoolPlugin has a button to send a message to MaltPipeline and some data, while MaltPipeline wants to tell CoolPlugin what is the histogram of the image or something (which we'll consider big enough data to use the shared buffer).
- Prior setup
   - I don't register CoolPlugin to blender, but I have registed Malt as usual with the setup
   - CoolPlugin is stored in the same folder as my MaltPipeline, 
- Startup
    - BlenderMalt starts, and will automatically find my plugin and start it once Malt is up.
    - CoolPlugin then will go through the registration process as usual, but started by BlenderMalt. It will also get the reference then.
- Operation: CoolPlugin to MaltPipeline
    - User presses the button
    - CoolPlugin uses its BlenderMalt reference to send a Custom message with a simple dict with `numberOfButtonPresses` and `coolMessageOfTheDay`
    - BlenderMalt will then execute a callback inside MaltPipeline, which can do whatever I need at that moment
- Operation: MaltPipeline to CoolPlugin
   - Some render happens and I want to send the histogram back.
   - I take MaltPipeline's BlenderMalt reference, and from that get access to the shared_buffer.
   - I put my histogram into the shared buffer.
   - I then use my BlenderMalt reference to send a simple message to CoolPlugin that basically says "hey your histogram is ready"
   - CoolPlugin is then called via a callback, and may then access to shared buffer to do whatever with it.

Some points I'm not sure about:
1. How exactly does BlenderMalt understand what is a plugin and what is a random python file ?
2. How does CoolPlugin get the BlenderMalt reference ?
3. In which circumstances will CoolPlugin be restarted ? From a quick read, it seems like it will be at the same time, but I'm not sure.
4. What does that look like on the Malt side ? I don't see any callback or reference to BlenderMalt in `Pipeline.py`

I probably missed a few implementation points, but how far am I of the mark ?

> You can certainly import a python module from both processes, and you can import any Malt module from BlenderMalt.
But it won't be THE SAME module, it's the same module loaded twice from different processes.

Okay then, that would be unsuitable for that usecase, since I would need the same instance otherwise it won't have access to the correct data.

---

> Sure, what I mean is that it's ok to open a pull request for a specific feature even if the code is not ready to be merged, or even if there's no code yet at all but you want to discuss the implementation and design details.

Ah okay haha, I've usually seen the workflow of discussion -> issue -> pull request used.

If I sum up the remaining issues how I see it at the moment, and what is needed before I/we start working:
- Buttons
    - I think the logic is somewhat sound here of having it only happen once (if you confirm)
    - We haven't talked about what form will the callback take. Maybe a BlenderMalt custom message too ?
    - We haven't talked about how to declare them either. Maybe as a special parameter ?
- UI control
    - Still in discussion.
    - Seems like grouping is already in but not folding, do you have an idea of where I could look to implement it ? (even if rough, just to point me in the correct direction)
- Interprocess communication
   - Still in discussion
   - Seems like it's already implemented, so maybe the task is just writing documentation / examples
   
Thanks for all the time you spend with me here btw :)

**Comment by pragma37** [comment]
> Then why would we want to call a function several times when pressing a button if there is only one instance ? I would think a single call is the simplest and expected behavior

Because:
>The issue is that each Viewport has its own Scene and Pipeline instance, so if you want to do something scene-related you would have to do it per Viewport.

It may not be needed for your specific use case since I guess all your scenes share the same materials (?), but that's not the usual case.
Think of a button for baking textures, or pre-computing light probes, for example.

> I didn't want to clog it with potentially unneeded info because it's a bit long haha, I'll try to keep it concise.
(...)

I think another (more simple) solution for your issue (and that still makes sense for the general use case) could be to add an option to fold material properties specifically, since they are the only parameter type that can increase the number of UI items arbitrarily.

We could still add the foldable categories too if you want, though.

> Maybe with remove_property ? I haven't evaluated this yet because I learnt about it when you told me lol, but it seems like it would also encounter some of the problems of 1. (sans the risk of not saving)

Beware, remove_property actually removes the property, it's not a UI thing. I mentioned it in the context of Malt remembering the values for undeclared parameters.

> so I could do GroupA.ParamX and GroupA.ParamY, and they would get grouped together automatically ?

Yep! It also works for uniforms.

> Okay, so I'll try to phrase how I understand it works, let me know if I have a mistake somewhere or if you have more detail

> CoolPlugin then will go through the registration process as usual, but started by BlenderMalt. It will also get the reference then.

There's no explicit reference, but since they're both part of the same process you can do something like:
`from BlenderMalt.MaltPipeline import get_bridge`

> BlenderMalt will then execute a callback inside MaltPipeline, which can do whatever I need at that moment

BlenderMalt doesn't have direct access to Malt, only to Bridge.

```mermaid
graph LR
    BM(BlenderMalt)
    C(Client_API.py)
    classDef Blender stroke:#fca503;
    class BM,C Blender;
    S(Server.py)
    M(Malt)
    classDef Malt stroke:#03fc4e;
    class S,M Malt;

    subgraph Blender Process
    subgraph P1 [ ]
        subgraph Bridge 
            C
        end
        BM <--> C
    end
    end
    C <--> S
    subgraph Malt Process
    subgraph P2 [ ]
        subgraph Bridge 
            S
        end
        S <--> M
    end
    end

    classDef Padding fill:none,stroke:none;
    class P1,P2 Padding
```

> How exactly does BlenderMalt understand what is a plugin and what is a random python file ?

It simply tries to import any folder in the plugin dir as a python module and tries to call PLUGIN.blendermalt_register.
https://github.com/bnpr/Malt/blob/Development/BlenderMalt/__init__.py#L161

> In which circumstances will CoolPlugin be restarted ? From a quick read, it seems like it will be at the same time, but I'm not sure.

Never, unless you explicitly disable/enable BlenderMalt or use the Blender Reload Scripts operator.

> What does that look like on the Malt side ? I don't see any callback or reference to BlenderMalt in Pipeline.py

Malt **can't** have access to BlenderMalt, they're two separate processes. All the communications are done through the Bridge interface.

> We haven't talked about what form will the callback take. Maybe a BlenderMalt custom message too ?
We haven't talked about how to declare them either. Maybe as a special parameter ?

Yes to both.

> Seems like grouping is already in but not folding, do you have an idea of where I could look to implement it ? (even if rough, just to point me in the correct direction)

Here's where the UI is drawn: https://github.com/bnpr/Malt/blob/9182cb2e69a2a7e4031b7a8ee69295859e9dadee/BlenderMalt/MaltProperties.py#L566
We should have a boolean property for each group that should be rendered in the UI along its label:
https://github.com/bnpr/Malt/blob/9182cb2e69a2a7e4031b7a8ee69295859e9dadee/BlenderMalt/MaltProperties.py#L636
In the `for key in keys` iteration that bool should be checked to know if the parameter must be rendered or not.

> Interprocess communication
>
>    Still in discussion
    Seems like it's already implemented, so maybe the task is just writing documentation / examples

I'd say, read Bridge.Client_API.py,  that should give you a good idea of how it works.

> Thanks for all the time you spend with me here btw :)

👍 👍 👍

**Comment by panthavma** [comment]
> It may not be needed for your specific use case since I guess all your scenes share the same materials (?), but that's not the usual case.
Think of a button for baking textures, or pre-computing light probes, for example.

Oh okay! I understand better now.

> I think another (more simple) solution for your issue (and that still makes sense for the general use case) could be to add an option to fold material properties specifically, since they are the only parameter type that can increase the number of UI items arbitrarily.
We could still add the foldable categories too if you want, though.

That's a nice idea! I think there's still value to the foldable categories, because even when empty the list is long, but both can work in tandem as you suggest.

> Beware, remove_property actually removes the property, it's not a UI thing. I mentioned it in the context of Malt remembering the values for undeclared parameters.

Oh okay, thanks for the clarification!

> There's no explicit reference, but since they're both part of the same process you can do something like:
from BlenderMalt.MaltPipeline import get_bridge

Oooh I didn't realize it was the same process, I understand that better now!

> Never, unless you explicitly disable/enable BlenderMalt or use the Blender Reload Scripts operator.

Oh okay, so that's the same as a regular plugin if I understand correctly ?

> Here's where the UI is drawn: [...]

Thanks !

> I'd say, read Bridge.Client_API.py, that should give you a good idea of how it works.

Neat! I'll do that.

---

I think I'll start with the folding, since that *seems* to have less ramifications and be more directly useful. I'll get a PR when it starts to look decent, or come back if I have more questions.

**Comment by panthavma** [comment]
Hello, to follow up on this, I've finally been able to work on this and managed to make *most* of foldable categories work
![image](https://github.com/bnpr/Malt/assets/22663538/4b1b1d15-74ef-456b-9e4f-5e1e6c30a87d)

I'll do a PR as you suggest when I can, as there are some issues still in my implementation that I'll need help with before merging.

**Reply by pragma37** [reply]
That looks nice! 😀
